jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java

```diff
@@ -0,0 +1,570 @@
+
+		Class<?>[] argTypes = targetMethodType.parameterArray();
+		int argTypeCount = argTypes.length;
+		argLayouts = funcDesc.argumentLayouts();
+		int argLayoutCount = argLayouts.size();
+		if (argTypeCount != argLayoutCount) {
+			throw new IllegalArgumentException("The arity (" + argTypeCount //$NON-NLS-1$
+												+ ") of the argument types is inconsistent with the arity ("  //$NON-NLS-1$
+												+ argLayoutCount + ") of the argument layouts");  //$NON-NLS-1$
+		}
+		
+		argLayoutArray = argLayouts.toArray(new MemoryLayout[argLayoutCount]);
+		for (int argIndex = 0; argIndex < argLayoutCount; argIndex++) {
+			if (!validateArgRetTypeClass(argTypes[argIndex])) {
+				throw new IllegalArgumentException("The passed-in argument type at index " + argIndex + " is neither primitive nor MemoryAddress"); //$NON-NLS-1$ //$NON-NLS-2$
+			}
+			validateLayoutAgainstType(argLayoutArray[argIndex], argTypes[argIndex]);
+		}
+	}
+	
+	/* Verify whether the specified type is primitive, MemoryAddress (for pointer) */
+	private static boolean validateArgRetTypeClass(Class<?> targetType) {
+		if (!targetType.isPrimitive() && (targetType != MemoryAddress.class)
+		) {
+			return false;
+		}
+		return true;
+	}
+	
+	/* Check the validity of the layout against the corresponding type */
+	private static void validateLayoutAgainstType(MemoryLayout targetLayout, Class<?> targetType) {
+		boolean isPrimitiveLayout = false;
+		
+		if (targetLayout != null) {
+			if (!targetLayout.hasSize()) {
+				throw new IllegalArgumentException("The layout's size is expected: layout = " + targetLayout); //$NON-NLS-1$
+			} else if (targetLayout.bitSize() <= 0) {
+				throw new IllegalArgumentException("The layout's size must be greater than zero: layout = " + targetLayout); //$NON-NLS-1$
+			}
+		}
+		
+		if (((targetType == void.class) && (targetLayout != null))
+		|| ((targetType != void.class) && (targetLayout == null))
+		) {
+			throw new IllegalArgumentException("Mismatch between the layout and the type: layout = "  //$NON-NLS-1$
+												+ ((targetLayout == null) ? "VOID" : targetLayout) //$NON-NLS-1$
+												+ ", type = " + targetType);  //$NON-NLS-1$
+		/* Check the primitive type and MemoryAddress against the ValueLayout */
+		} else if (targetType != void.class) {
+			if (!ValueLayout.class.isInstance(targetLayout)) {
+				throw new IllegalArgumentException("ValueLayout is expected: layout = " + targetLayout); //$NON-NLS-1$
+			}
+			/* Check the size and kind of the ValueLayout for the primitive types and MemoryAddress */
+			validateValueLayoutSize(targetLayout, targetType);
+			validateValueLayoutKind(targetLayout, targetType);
+		}
+	}
+	
+	/* Check the size of the specified primitive layout to determine whether it matches the specified type */
+	private static void validateValueLayoutSize(MemoryLayout TypeLayout, Class<?> targetType) {
+		String layoutSize = TypeLayout.bitSize() + "_bits"; //$NON-NLS-1$
```

Comment:
This seems a rather expensive way to handle certain values (8, 16, 32, 64) of a `long`.

Reply:
Agreed and replaced with directly integer to check the bit size.